### DISCUSSION TITLE
Docker container with CUDA and ifm3d

### DISCUSSION THREAD

**Discussion by j-friedl** [ifm/ifm3d#360] [Q&A] [answered]
Hello again, it seems I'm a regular customer here ;)

I'm wondering if there's a better way to use your Docker containers. As it's hard to use the `ghcr.io/ifm/ifm3d:stable` image as base (no root privileges) and it doesn't seem to support CUDA I've written my own Docker Image based upon `nvcr.io/nvidia/l4t-base:r32.4.3`. However, the installation if the ifm3d software seems tedious inside the container. Also the image size is growing rapidly. What would be the workflow to develop a C++ application with your ifm3d library and CUDA support inside a docker container? My current Dockerfile looks something like this: (Note that it is not multistage rn, but if there's no better solution, I'll have to implement a multistage buil)

```
# Base linux for tegra (l4t) amr64/aarch64 image (see https://catalog.ngc.nvidia.com/orgs/nvidia/containers/l4t-base)
FROM nvcr.io/nvidia/l4t-base:r32.4.3 AS build

# Update system
RUN apt-get update && apt-get -y upgrade

# Set timezone for tzdata docker-build bug (see https://grigorkh.medium.com/fix-tzdata-hangs-docker-image-build-cdb52cc3360d)
ENV TZ=Europe/Vienna
RUN ln -snf /usr/share/zoneinfo/$TZ /etc/localtime && echo $TZ > /etc/timezone

# Install required apt packages
RUN apt-get install -y --no-install-recommends \
    git \
    jq \
    libcurl4-openssl-dev \
    libgtest-dev \
    libgoogle-glog-dev \
    libxmlrpc-c++8-dev \
    libopencv-dev \
    libpcl-dev \
    libproj-dev \
    build-essential \
    coreutils \
    curl \
    ca-certificates \
    gnupg

# Update CMake to version 3.23.2 (needed for IFM software)
ADD https://cmake.org/files/v3.23/cmake-3.23.2.tar.gz /opt
RUN mkdir /opt/cmake-3.23.2 &&\
    tar -xf /opt/cmake-3.23.2.tar.gz -C /opt &&\
    cd /opt/cmake-3.23.2 &&\
    ./configure &&\
    make install

# Install IFM libraries (with workaround)
WORKDIR /home/oem/nvidia-container-sample/Workspace/external
ADD ./external/ifm3d-0.93.0_working.tar.xz .
RUN mkdir ifm3d-0.93.0/build &&\
    cd /home/oem/nvidia-container-sample/Workspace/external/ifm3d-0.93.0/build &&\
    cmake .. &&\
    cmake --build .

# Copy local files to container
WORKDIR /home/oem
COPY ./Workspace nvidia-container-sample/Workspace

# Build files
# TODO: make and build files

# Multistage image
#FROM nvcr.io/nvidia/l4t-base:r32.4.3 AS base
#COPY <builded app>
#CMD[./app]
```

**Comment by graugans** [comment, marked as answer]
Hej @j-friedl , the example Docker containers are a showcase and will not be able to fit all use-cases. You are right the `aarch64` Docker host emulation is quite slow. We are internally discussing how we cloud provide pre build binaries to accelerate this.

When it comes to a size reduction of the Docker images the multistage build is the way to go. As a small enhancement you can download a pre-built CMake version from the CMake project.

**Reply by j-friedl** [reply]
"small enhancement" is good. The compilation of CMake took ages, thank you for the hint! :)
